edb/edgeql/compiler/context.py

```diff
@@ -569,6 +569,28 @@ class ContextLevel(compiler.ContextLevel):
     active_defaults: FrozenSet[s_objtypes.ObjectType]
     """For detecting cycles in defaults"""
 
+    collection_cast_path: Optional[list[Tuple[str, Optional[str]]]]
+    """For generating errors messages when casting to collections.
+        Only ever set in outermost cast.
+
+    The cast path element is a tuple of the collection type and an optional
+    element name. eg. ('tuple', 'a') or ('array', None)
+
+    The list is shared between the outermost context and all its sub contexts.
+    When casting a collection, each element's path should be pushed before
+    entering the "sub-cast" and popped immediately after.
+
+    In the event of a cast error, the list is preserved at the outermost cast.
+    """
+
+    collection_cast_from_type: Optional[s_types.Type]
+    """For generating errors messages when casting to collections.
+        Only ever set in outermost cast."""
+
+    collection_cast_to_type: Optional[s_types.Type]
+    """For generating errors messages when casting to collections.
+        Only ever set in outermost cast."""
```

Comment:
Maybe this would be nicer if we moved it all to one `CollectionCastInfo` object?